File: core/domain/service/Ordering.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

// Package service holds the domain services: the rules that belong to no single entity.
//
// They are pure functions over values. Nothing here reads a clock, a database, or a request -
// which is what makes the rules testable at their boundaries rather than through three layers of
// fakes.
package service

import (
	"strings"

	"github.com/Jersyfi/hubtask/core/domain/model/shared"
)

// Rank keys, or "fractional indexing" (offline-sync.md §4.2).
//
// A position is a lexicographic key between its neighbours rather than an integer. Inserting
// between two rows writes one row instead of renumbering every successor - which is what lets two
// offline devices insert into the same list without either one's order being discarded on sync
// (test SY-4).
//
// The scheme is the established one described by David Greenspan ("Implementing Fractional
// Indexing", 2021) and implemented by the `fractional-indexing` library: a key is an integer part
// followed by an optional fraction, and the integer part's first character encodes how many digits
// follow it. That encoding is what keeps appending cheap. Without it, every append would have to
// sit lexicographically inside the previous key's space, and a thousand appends would produce a
// thousand-character key; with it, `a0` is followed by `a1`, `az` by `b00`, and the length grows
// with the logarithm of the number of rows.
//
// Letters `a`..`z` head the non-negative integers and `A`..`Z` the negative ones, so a key can
// always be produced *before* the first row as well - which is what "move to the top" needs.

// orderKeyDigits are the digits of a key, in ASCII order, so that comparing two keys as strings
// compares them as numbers. Base 62 keeps a key short, and a database sorts it with a plain index.
const orderKeyDigits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

// orderKeyZero is the first key of an empty list: integer zero, no fraction.
const orderKeyZero = "a0"

// smallestOrderInteger is the lower end of the integer space. A key may not be exactly this
// value, because there would then be no key left below it and "move to the top" would have no
// answer.
var smallestOrderInteger = "A" + strings.Repeat("0", 26)

// OrderKeyBetween returns a key that sorts strictly after previous and strictly before next.
//
// An empty previous means "before everything", an empty next means "after everything". Both empty
// gives the first key of an empty list. A key is never empty, so the empty string is unambiguous
// as "no bound".
func OrderKeyBetween(previous, next string) (string, error) {
	if err := validateOrderKey(previous); err != nil {
		return "", err
	}
	if err := validateOrderKey(next); err != nil {
		return "", err
	}
	if previous != "" && next != "" && previous >= next {
		// The caller passed neighbours that are not neighbours, or passed them the wrong way
		// round. Nothing a client sent could have caused it (security.md §9).
		return "", shared.ErrInternal.
			WithDetail("ordering.bounds_invalid").
			WithParams(map[string]string{"previous": previous, "next": next})
	}

	switch {
	case previous == "" && next == "":
		return orderKeyZero, nil
	case previous == "":
		return keyBefore(next)
	case next == "":
		return keyAfter(previous)
	}
	return keyBetween(previous, next)
}

// OrderKeyAfter returns a key that sorts after previous. It is what appending to a list needs: an
// empty previous is an empty list.
func OrderKeyAfter(previous string) (string, error) { return OrderKeyBetween(previous, "") }

// keyBefore is the "move to the top" case: one integer step below next, or half of next's
// fraction when there is no integer step left.
func keyBefore(next string) (string, error) {
	integer, fraction := splitOrderKey(next)

	if integer == smallestOrderInteger {
		middle, err := midpoint("", fraction)
		return integer + middle, err
	}
	// next carries a fraction, so its own integer part already sorts below it.
	if integer < next {
		return integer, nil
	}

	decremented, err := decrementInteger(integer)
	if err != nil {
		return "", err
	}
	if decremented == "" {
		return "", shared.ErrInternal.
			WithDetail("ordering.space_exhausted").
			WithParams(map[string]string{"next": next})
	}
	return decremented, nil
}

// keyAfter is the append case, and the one that has to stay short: one integer step up, and only
// when the integer space is exhausted does the key grow a fraction.
func keyAfter(previous string) (string, error) {
	integer, fraction := splitOrderKey(previous)

	incremented, err := incrementInteger(integer)
	if err != nil {
		return "", err
	}
	if incremented != "" {
		return incremented, nil
	}

	middle, err := midpoint(fraction, "")
	return integer + middle, err
}

func keyBetween(previous, next string) (string, error) {
	previousInteger, previousFraction := splitOrderKey(previous)
	nextInteger, nextFraction := splitOrderKey(next)

	if previousInteger == nextInteger {
		middle, err := midpoint(previousFraction, nextFraction)
		return previousInteger + middle, err
	}

	incremented, err := incrementInteger(previousInteger)
	if err != nil {
		return "", err
	}
	// One integer step up is the shortest key there is, as long as it still sorts below next.
	if incremented != "" && incremented < next {
		return incremented, nil
	}

	middle, err := midpoint(previousFraction, "")
	return previousInteger + middle, err
}

// midpoint returns the shortest fraction strictly between a and b, where an empty b means "1".
// Both are fractions without a leading dot and without a trailing zero digit.
func midpoint(a, b string) (string, error) {
	zero := orderKeyDigits[0]

	if b != "" && a >= b {
		return "", shared.ErrInternal.
			WithDetail("ordering.bounds_invalid").
			WithParams(map[string]string{"previous": a, "next": b})
	}
	if strings.HasSuffix(a, string(zero)) || (b != "" && strings.HasSuffix(b, string(zero))) {
		// A trailing zero is a fraction with a shorter equal - it would break the comparison the
		// whole scheme rests on.
		return "", shared.ErrInternal.
			WithDetail("ordering.key_malformed").
			WithParams(map[string]string{"key": a + " " + b})
	}

	if b != "" {
		// A shared prefix is carried over untouched; only the first differing digit decides.
		common := 0
		for common < len(b) {
			digit := zero
			if common < len(a) {
				digit = a[common]
			}
			if digit != b[common] {
				break
			}
			common++
		}
		if common > 0 {
			rest, err := midpoint(cut(a, common), b[common:])
			return b[:common] + rest, err
		}
	}

	lower := 0
	if a != "" {
		lower = digitIndex(a[0])
	}
	upper := len(orderKeyDigits)
	if b != "" {
		upper = digitIndex(b[0])
	}

	if upper-lower > 1 {
		// Rounded rather than truncated, so that the digit sits in the middle of the gap and the
		// next insertion on either side has the same room.
		return string(orderKeyDigits[(lower+upper+1)/2]), nil
	}
	if len(b) > 1 {
		return b[:1], nil
	}
	// The digits are adjacent and b has nothing more to give: keep a's digit and look one
	// position further, where b no longer constrains anything.
	rest, err := midpoint(cut(a, 1), "")
	return string(orderKeyDigits[lower]) + rest, err
}

// cut is s[n:] for a string that may be shorter than n.
func cut(s string, n int) string {
	if n >= len(s) {
		return ""
	}
	return s[n:]
}

// integerLength reads the length the head character declares. It is the trick that makes appending
// cheap: `b00` sorts after `az` because `b` follows `a`, not because it is longer.
func integerLength(head byte) (int, error) {
	switch {
	case head >= 'a' && head <= 'z':
		return int(head-'a') + 2, nil
	case head >= 'A' && head <= 'Z':
		return int('Z'-head) + 2, nil
	}
	return 0, shared.ErrInternal.
		WithDetail("ordering.key_malformed").
		WithParams(map[string]string{"key": string(head)})
}

// splitOrderKey separates the integer part from the fraction.
//
// A key whose head declares more digits than it has is returned whole, so that validateInteger
// gets to report it rather than this function reaching past the end of the string.
func splitOrderKey(key string) (integer, fraction string) {
	length, err := integerLength(key[0])
	if err != nil || length > len(key) {
		return key, ""
	}
	return key[:length], key[length:]
}

// incrementInteger returns the next integer, or the empty string when there is none - the top of
// the space, where the caller falls back to a fraction.
func incrementInteger(integer string) (string, error) {
	if err := validateInteger(integer); err != nil {
		return "", err
	}

	head, digits := integer[0], []byte(integer[1:])
	carry := true
	for i := len(digits) - 1; carry && i >= 0; i-- {
		if next := digitIndex(digits[i]) + 1; next == len(orderKeyDigits) {
			digits[i] = orderKeyDigits[0]
		} else {
			digits[i] = orderKeyDigits[next]
			carry = false
		}
	}

	if !carry {
		return string(head) + string(digits), nil
	}
	switch head {
	case 'Z':
		// The last negative integer is followed by zero.
		return orderKeyZero, nil
	case 'z':
		return "", nil
	}
	if head+1 > 'a' {
		digits = append(digits, orderKeyDigits[0])
	} else {
		digits = digits[1:]
	}
	return string(head+1) + string(digits), nil
}

// decrementInteger returns the previous integer, or the empty string at the bottom of the space.
func decrementInteger(integer string) (string, error) {
	if err := validateInteger(integer); err != nil {
		return "", err
	}

	last := orderKeyDigits[len(orderKeyDigits)-1]
	head, digits := integer[0], []byte(integer[1:])
	borrow := true
	for i := len(digits) - 1; borrow && i >= 0; i-- {
		if previous := digitIndex(digits[i]) - 1; previous < 0 {
			digits[i] = last
		} else {
			digits[i] = orderKeyDigits[previous]
			borrow = false
		}
	}

	if !borrow {
		return string(head) + string(digits), nil
	}
	switch head {
	case 'a':
		return string('Z') + string(last), nil
	case 'A':
		return "", nil
	}
	if head-1 < 'Z' {
		digits = append(digits, last)
	} else {
		digits = digits[1:]
	}
	return string(head-1) + string(digits), nil
}

func validateInteger(integer string) error {
	if integer == "" {
		return shared.ErrInternal.WithDetail("ordering.key_malformed").
			WithParams(map[string]string{"key": integer})
	}
	length, err := integerLength(integer[0])
	if err != nil {
		return err
	}
	if len(integer) != length {
		return shared.ErrInternal.
			WithDetail("ordering.key_malformed").
			WithParams(map[string]string{"key": integer})
	}
	return nil
}

// validateOrderKey refuses a key this package could not have produced. The keys come from storage,
// so an unusable one means a row was written by something that does not share this scheme - and
// continuing would place the new key in an unpredictable position. The empty string is the absent
// bound and passes.
func validateOrderKey(key string) error {
	if key == "" {
		return nil
	}
	malformed := shared.ErrInternal.
		WithDetail("ordering.key_malformed").
		WithParams(map[string]string{"key": key})

	for i := range len(key) {
		if digitIndex(key[i]) < 0 {
			return malformed
		}
	}
	if key == smallestOrderInteger {
		// Permitted as a prefix, refused on its own: a key there would leave nothing below it.
		return malformed
	}

	integer, fraction := splitOrderKey(key)
	if err := validateInteger(integer); err != nil {
		return err
	}
	if strings.HasSuffix(fraction, string(orderKeyDigits[0])) {
		return malformed
	}
	return nil
}

func digitIndex(digit byte) int { return strings.IndexByte(orderKeyDigits, digit) }

File: packages/sync-engine/src/ordering.ts
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

// Fractional order keys, the server's scheme on this side of the wire
// (`core/domain/service/Ordering.go`, offline-sync.md §4.2).
//
// A reorder made offline has to name the rank it wants as a key rather than as "before that
// entry", because by the time it is pushed that entry may have moved; the key is a fractional
// index, so two devices inserting into one list keep both insertions. The scheme is David
// Greenspan's ("Implementing Fractional Indexing", 2021): an integer part whose first character
// encodes how many digits follow it, then an optional fraction, in base 62 digits in ASCII order,
// so that comparing two keys as strings compares them as numbers. This is the server's code,
// branch for branch, with the server's own tests ported: a key this mints is a key the server
// would have minted between the same neighbours.

const DIGITS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
const ZERO = DIGITS[0] as string;
const LAST = DIGITS[DIGITS.length - 1] as string;
/** The first key of an empty list: integer zero, no fraction. */
export const ORDER_KEY_ZERO = 'a0';
/** The lower end of the integer space; permitted as a prefix, never as a key of its own. */
const SMALLEST_INTEGER = `A${ZERO.repeat(26)}`;

export class OrderKeyError extends Error {
  readonly detailCode: string;

  constructor(detailCode: string, message: string) {
    super(message);
    this.name = 'OrderKeyError';
    this.detailCode = detailCode;
  }
}

const digitIndex = (digit: string): number => DIGITS.indexOf(digit);

/** The length the head character declares: `a`..`z` head the non-negative integers, `A`..`Z` the negative. */
function integerLength(head: string): number {
  if (head >= 'a' && head <= 'z') return head.charCodeAt(0) - 'a'.charCodeAt(0) + 2;
  if (head >= 'A' && head <= 'Z') return 'Z'.charCodeAt(0) - head.charCodeAt(0) + 2;
  throw new OrderKeyError('ordering.key_malformed', `not an order key head: ${head}`);
}

function splitKey(key: string): [integer: string, fraction: string] {
  let length: number;
  try {
    length = integerLength(key[0] as string);
  } catch {
    return [key, ''];
  }
  if (length > key.length) return [key, ''];
  return [key.slice(0, length), key.slice(length)];
}

function validateInteger(integer: string): void {
  if (integer === '' || integerLength(integer[0] as string) !== integer.length) {
    throw new OrderKeyError('ordering.key_malformed', `malformed integer part: ${integer}`);
  }
}

export function validateOrderKey(key: string): void {
  if (key === '') return;
  for (const digit of key) {
    if (digitIndex(digit) < 0) throw new OrderKeyError('ordering.key_malformed', `malformed key: ${key}`);
  }
  if (key === SMALLEST_INTEGER) throw new OrderKeyError('ordering.key_malformed', `malformed key: ${key}`);
  const [integer, fraction] = splitKey(key);
  validateInteger(integer);
  if (fraction.endsWith(ZERO)) throw new OrderKeyError('ordering.key_malformed', `malformed key: ${key}`);
}

/** The next integer, or the empty string at the top of the space. */
function incrementInteger(integer: string): string {
  validateInteger(integer);
  const head = integer[0] as string;
  const digits = integer.slice(1).split('');
  let carry = true;
  for (let i = digits.length - 1; carry && i >= 0; i -= 1) {
    const next = digitIndex(digits[i] as string) + 1;
    if (next === DIGITS.length) {
      digits[i] = ZERO;
    } else {
      digits[i] = DIGITS[next] as string;
      carry = false;
    }
  }
  if (!carry) return head + digits.join('');
  if (head === 'Z') return ORDER_KEY_ZERO;
  if (head === 'z') return '';
  const nextHead = String.fromCharCode(head.charCodeAt(0) + 1);
  if (nextHead > 'a') digits.push(ZERO);
  else digits.shift();
  return nextHead + digits.join('');
}

/** The previous integer, or the empty string at the bottom of the space. */
function decrementInteger(integer: string): string {
  validateInteger(integer);
  const head = integer[0] as string;
  const digits = integer.slice(1).split('');
  let borrow = true;
  for (let i = digits.length - 1; borrow && i >= 0; i -= 1) {
    const previous = digitIndex(digits[i] as string) - 1;
    if (previous < 0) {
      digits[i] = LAST;
    } else {
      digits[i] = DIGITS[previous] as string;
      borrow = false;
    }
  }
  if (!borrow) return head + digits.join('');
  if (head === 'a') return `Z${LAST}`;
  if (head === 'A') return '';
  const previousHead = String.fromCharCode(head.charCodeAt(0) - 1);
  if (previousHead < 'Z') digits.push(LAST);
  else digits.shift();
  return previousHead + digits.join('');
}

const cut = (s: string, n: number): string => (n >= s.length ? '' : s.slice(n));

/** The shortest fraction strictly between a and b, where an empty b means "1". */
function midpoint(a: string, b: string): string {
  if (b !== '' && a >= b) throw new OrderKeyError('ordering.bounds_invalid', `${a} is not below ${b}`);
  if (a.endsWith(ZERO) || (b !== '' && b.endsWith(ZERO))) {
    throw new OrderKeyError('ordering.key_malformed', `trailing zero: ${a} ${b}`);
  }
  if (b !== '') {
    let common = 0;
    while (common < b.length) {
      const digit = common < a.length ? (a[common] as string) : ZERO;
      if (digit !== b[common]) break;
      common += 1;
    }
    if (common > 0) return b.slice(0, common) + midpoint(cut(a, common), b.slice(common));
  }
  const lower = a !== '' ? digitIndex(a[0] as string) : 0;
  const upper = b !== '' ? digitIndex(b[0] as string) : DIGITS.length;
  if (upper - lower > 1) return DIGITS[Math.floor((lower + upper + 1) / 2)] as string;
  if (b.length > 1) return b.slice(0, 1);
  return (DIGITS[lower] as string) + midpoint(cut(a, 1), '');
}

function keyBefore(next: string): string {
  const [integer, fraction] = splitKey(next);
  if (integer === SMALLEST_INTEGER) return integer + midpoint('', fraction);
  if (integer < next) return integer;
  const decremented = decrementInteger(integer);
  if (decremented === '') throw new OrderKeyError('ordering.space_exhausted', `nothing below ${next}`);
  return decremented;
}

function keyAfter(previous: string): string {
  const [integer, fraction] = splitKey(previous);
  const incremented = incrementInteger(integer);
  if (incremented !== '') return incremented;
  return integer + midpoint(fraction, '');
}

function keyBetween(previous: string, next: string): string {
  const [previousInteger, previousFraction] = splitKey(previous);
  const [nextInteger, nextFraction] = splitKey(next);
  if (previousInteger === nextInteger) return previousInteger + midpoint(previousFraction, nextFraction);
  const incremented = incrementInteger(previousInteger);
  if (incremented !== '' && incremented < next) return incremented;
  return previousInteger + midpoint(previousFraction, '');
}

/**
 * A key that sorts strictly after `previous` and strictly before `next`. An empty `previous`
 * means "before everything", an empty `next` "after everything"; both empty is the first key of
 * an empty list.
 */
export function orderKeyBetween(previous: string, next: string): string {
  validateOrderKey(previous);
  validateOrderKey(next);
  if (previous !== '' && next !== '' && previous >= next) {
    throw new OrderKeyError('ordering.bounds_invalid', `${previous} is not below ${next}`);
  }
  if (previous === '' && next === '') return ORDER_KEY_ZERO;
  if (previous === '') return keyBefore(next);
  if (next === '') return keyAfter(previous);
  return keyBetween(previous, next);
}

export const orderKeyAfter = (previous: string): string => orderKeyBetween(previous, '');
